Bug 13711 - An Open angle bracket < in custom field within Manager Form crashes the save process

Finetuned opened the pull request:
### What does it do?
This is a two-part bugfix:

1. Encloses json output from the modConnectorResponse class in a <textarea> tag
2. Overrides the handleResponse function in Ext.data.Connection to remove the <textarea> tag from the response body after it has been processed by doFormUpload function in Ext.data.Connection

### Why is it needed?
A single angle bracket in a custom field in the Manager crashes the save process. This is due to Ext.data.Connection using a fake ajax response (?!):  Json returned from the server is dumped into the body of a hidden iframe. The browser truncates the json at the unescaped angle bracket triggering a json decode exception down the line.

The textarea encapsulation is already implemented in Ext.data.connection in doFormUpload.
[See the function cb for details.](http://docs.sencha.com/extjs/3.4.0/source/Connection.html#Ext-data-Connection)


### Related issue(s)/PR(s)
As listed in [13711](https://github.com/modxcms/revolution/issues/13711)

Finetuned committed:
enclosed json output in textarea tag to prevent the browser truncating unmatched symbols e.g. < when the response is dumped into an iframe body on return.

Finetuned committed:
Added an override to handleResponse() in Ext.data.Connection.
This removes enclosing textarea tags in the response body if they exist.

Side effect: needed to substitute the variable REQUESTCOMPLETE with its value 'requestcomplete' . This var is declared in the anonymous function encapsulating Ext.data.conection as it's not present in the override.

Mark-H commented on core/model/modx/modconnectorresponse.class.php:
```diff
@@ -179,7 +179,7 @@ public function outputContent(array $options = array()) {
                 $json = $modx->stripTags($_GET['callback']) . '(' . $json . ')';
             }
-            die($json);
+            die('<textarea>'. $json .'</textarea>');
```
I remember you talking about a hack-ish approach, but I didn't quite see this one coming 😬This is going to break interaction with connectors from non-extjs sources, which I know people are doing, so I think we'll have to go back to the drawing board on this one...

Finetuned replied:
Yes, I thought this might have far-reaching consequences. This was working back from Ext.data.Connection which provides for the textarea addition

Mark-H commented on manager/assets/modext/util/utilities.js:
```diff
@@ -319,6 +319,26 @@ Ext.override(Ext.form.Action.Submit,{
+        var k = j.argument.options;
+        j.argument = k ? k.argument : null;
+        // trim out the enclosing textarea tag (the long way to avoid any tags in the response body) if exists
+        j.responseText = j.responseText.replace('<textarea>', '');
```
I suspect this would also remove textarea's from within the content (say, someone having a form markup in a chunks' content). 

Is there a different way to handle `j.responseText` here to fix the issue, which doesn't require the `<textarea>` markup to be returned from the processor? Maybe that can be dynamically added in this handleResponse method or elsewhere to fix the issue?

Finetuned replied:
Noted. I added this in to restore 'normal behaviour ' after the enclosing above. I could be more specific about the stripping but it doesn't resolve the first problem. Back to square one then : s

Mark-H reviewed (changes requested):
Please see the code comments - this will break too much to merge in its current state I'm afraid :(

Finetuned committed:
reverted addition of <textarea> wrapper

Finetuned committed:
Removed the handleResponse() override in Ext.data.Connection

Finetuned committed:
Added an override for Ext.data.connection  doFormUpload()

Variables defined in the closure in which Ext.data.Connection is declared are not available in the override and need to be redeclared.

Finetuned committed:
Added a new property fallbackResponse to Ext.dataConnection:
This object emulates the structure of the json object retuend by the server.

Added a new function isValidResponse() tp Ext.data.Connection:
This function tests whether the passed in response object's responseText is valid JSON.
It returns the original stringified object if  valid.
It returns the stringified version fallbackResponse object if the result is false.

The message property of fallbackResponse is populated with an error message that should be replaced with a new entry in the lexicon

Finetuned committed:
Added a call to the function isValidResponse() in the nested cb() function to ensure a valid json object is passed along to handleResponse and so on.

Finetuned committed:
Added a new function isValidJSON to Ext.util.JSON to avoid inlining the code in isValidResponse() in Ext.data.Connection

Mark-H commented:
I may have found a better fix for this, see https://github.com/modxcms/revolution/issues/13711#issuecomment-401204379

Mark-H commented:
Closing as this would introduce non-acceptable breakage. The other proposal might work better but has not yet been turned into a PR.
